File: src/main/java/me/ksviety/teleporter/providers/SafePositionProvider.java
package me.ksviety.teleporter.providers;

import me.ksviety.teleporter.IPositionProvider;
import me.ksviety.teleporter.ShiftAxis;
import me.ksviety.teleporter.ShiftedVec3i;
import me.ksviety.teleporter.exceptions.CannotFindClosetSafePositionException;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3i;
import net.minecraft.world.World;

import java.util.Collection;
import java.util.Objects;

/**
 * Can throw <a>CannotFindClosestSafePositionException</a>
 */
public class SafePositionProvider implements IPositionProvider {
    private final static ShiftAxis[] SHIFT_DIRECTIONS = new ShiftAxis[] {
            ShiftAxis.Forward,
            ShiftAxis.Backward,
            ShiftAxis.Right,
            ShiftAxis.Left
    };

    private final IPositionProvider positionProvider;
    private final World world;
    private final Collection<String> bannedBlocks;
    private final int shiftRadius;
    private final int maxSearchIterations;

    public SafePositionProvider(
            IPositionProvider positionProvider,
            World world,
            Collection<String> bannedBlocks,
            int shiftRadius,
            int maxSearchIterations
    ) {
        this.positionProvider = positionProvider;
        this.world = world;
        this.bannedBlocks = bannedBlocks;
        this.shiftRadius = shiftRadius;
        this.maxSearchIterations = maxSearchIterations;
    }

    @Override
    public Vec3i provide() {
        return getSafePosition();
    }

    private Vec3i getSafePosition(int iteration) {
        final Vec3i position = positionProvider.provide();

        for (int shift = 0; shift < shiftRadius; shift++) {
            for (ShiftAxis axis : SHIFT_DIRECTIONS) {
                final Vec3i topBlockPosition = world.getTopSolidOrLiquidBlock(
                        new BlockPos(
                                ShiftedVec3i.shift(position, shift, axis)
                        )
                );

                // Shift up by one to avoid spawning inside the block
                final Vec3i shiftedPosition = ShiftedVec3i.shift(topBlockPosition, 1, ShiftAxis.Up);

                if (isSafe(shiftedPosition))
                    return shiftedPosition;
            }
        }

        if (iteration >= maxSearchIterations)
            throw new CannotFindClosetSafePositionException();

        return getSafePosition(iteration + 1);
    }

    private Vec3i getSafePosition() {
        return getSafePosition(0);
    }

    private boolean isSafe(Vec3i position) {
        try {
            final BlockPos positionBelowPlayer = new BlockPos(
                    ShiftedVec3i.shift(position, 1, ShiftAxis.Down)
            );

            if (world.isAirBlock(positionBelowPlayer))
                return false;

            final String name = Objects.requireNonNull(
                    world.getBlockState(positionBelowPlayer).getBlock().getRegistryName()
            ).getResourcePath();

            if (bannedBlocks.contains(name))
                return false;
        } catch (NullPointerException npe) {
            return false;
        }

        return true;
    }
}

File: src/main/kotlin/me/ksviety/teleporter/providers/SafePositionProvider.kt
package me.ksviety.teleporter.providers

import me.ksviety.teleporter.PositionProvider
import net.minecraft.world.World
import net.minecraft.util.math.Vec3i
import me.ksviety.teleporter.ShiftAxis
import net.minecraft.util.math.BlockPos
import me.ksviety.teleporter.ShiftedVec3i
import me.ksviety.teleporter.exceptions.CannotFindClosestSafePositionException

/**
 * Can throw <a>CannotFindClosestSafePositionException</a>
 */
class SafePositionProvider(
    private val positionProvider: PositionProvider,
    private val world: World,
    private val bannedBlocks: Collection<String>,
    private val shiftRadius: Int,
    private val maxSearchIterations: Int
) : PositionProvider {
    private val Vec3i.isSafe: Boolean
        get() {
            val positionBelowPlayer = BlockPos(
                ShiftedVec3i.shift(this, 1, ShiftAxis.Down)
            )

            if (world.isAirBlock(positionBelowPlayer))
                return false

            val name = world.getBlockState(positionBelowPlayer).block.registryName?.resourcePath ?: return false

            if (bannedBlocks.contains(name))
                return false

            return true
        }

    override fun provide(): Vec3i {
        return getSafePosition()
    }

    private fun getSafePosition(iteration: Int): Vec3i {
        val position = positionProvider.provide()

        for (shift in 0 until shiftRadius) {
            for (axis in SHIFT_DIRECTIONS) {
                val topBlockPosition = world.getTopSolidOrLiquidBlock(
                    BlockPos(
                        ShiftedVec3i.shift(position, shift, axis)
                    )
                )

                // Shift up by one to avoid spawning inside the block
                val shiftedPosition = ShiftedVec3i.shift(topBlockPosition, 1, ShiftAxis.Up)

                if (shiftedPosition.isSafe)
                    return shiftedPosition
            }
        }

        if (iteration >= maxSearchIterations)
            throw CannotFindClosestSafePositionException

        return getSafePosition(iteration + 1)
    }

    private fun getSafePosition() = getSafePosition(0)

    companion object {
        private val SHIFT_DIRECTIONS = arrayOf(
            ShiftAxis.Forward,
            ShiftAxis.Backward,
            ShiftAxis.Right,
            ShiftAxis.Left
        )
    }
}
